pmlc/rt/vulkan/vulkan_runtime.cc

```diff
@@ -719,3 +722,97 @@ LogicalResult VulkanRuntime::updateHostMemoryBuffers() {
   }
   return success();
 }
+
+struct Action {
+  virtual ~Action() {}
+};
+
+using ActionPtr = std::shared_ptr<Action>;
+
+struct LaunchKernelAction : Action {
+  VkPipeline pipeline;
+  VkPipelineLayout pipelineLayout;
+  SmallVector<VkDescriptorSet, 4> descriptorSets;
+  NumWorkGroups workGroups;
+  SmallVector<VkBufferMemoryBarrier, 4> deps;
+};
+
+struct MemoryTransferAction : Action {
+  VkBuffer src;
+  VkBuffer dst;
+  SmallVector<VkBufferCopy, 1> regions;
+};
+
+LogicalResult VulkanRuntime::createSchedule() {
+  std::vector<ActionPtr> schedule;
```

Comment:
TODO: this needs to be filled in. We need to figure out a way to turn the host-side MLIR module into this schedule.

For each user input and output (the arguments to the main entrypoint of the program), we'd need to create a `MemoryTransferAction` that represents a copy from host memory to device memory (in the case of inputs) and a copy from device memory to host memory (in the case of outputs).

For each `gpu.launch_func`, we'd want to create a `LaunchKernelAction`. This is basically a tuple of: `vector<memref>` representing the memrefs that a kernel depends on (either they are user inputs or they are the outputs of other kernels). The `pipeline`, `pipelineLayout`, `descriptorSets`, and `workGroups` are all based off of the `gpu.launch_func` op itself. Each of the memrefs that a kernel uses should turn into an entry in the `deps` field.

Reply:
Thanks. So we need to convert each `gpu.launch_func` to something like `launch_kernel(memref_deps)` in which we call `MemoryTransferAction` and `VulkanRuntime::createSchedule()`. (Please correct me if I misunderstood.) I think in this case we can add a new pass after gpu kernel outlining pass to do this work instead of making our own outlining pass.

Reply:
so the host-side MLIR will be turned into  calls to VulkanRuntime::createSchedule() .

Reply:
Or perhaps calls to something that builds up a schedule so that at the very end we say `runOnVulkan` which internally calls `createSchedule`.